### PR TITLE
named nodes

### PR DESCRIPTION
**Pull request by PythonFZ**
Had to make a new PR, because of rebasing conflicts on the other one.

# TODOs
- [x] add documentation

# Example

A Node can now be named, this allows using the same Node in different dependencies or using the same Node multiple times with slightly different parameters. E.g. this can be useful for adding multiple files. The file adding syntax does not change, and all Nodes are required to be on the same graph.

```py
    HelloWorld()(inputs=3)
    HelloWorld(name="Test01")(inputs=17)
    HelloWorld(name="Test02")(inputs=42)

    project.repro()

    assert HelloWorld(load=True).output == 3
    assert HelloWorld(load=True, name="Test01").output == 17
    assert HelloWorld(load=True, name="Test02").output == 42
```

### CODE REVIEW

File: zntrack/core/zntrack.py [hunk 1]
```diff
@@ -94,11 +94,23 @@ def __init__(self, child):
...
+        name: str
+            Custom name for the stage/Node that will overwrite the autogenerated
+            name based on the class.__name__
+        load: bool
+            set the stage to be loaded
+        """
+        self.stage_name = name
+        self.load = load
+        self.has_metadata = has_metadata
         pass
```
Comment (SamTov) [inline zntrack/core/zntrack.py:114]: why the pass?

File: zntrack/core/zntrack.py [hunk 2]
```diff
@@ -94,11 +94,23 @@ def __init__(self, child):
     @property
     def zn_files(self) -> ZnFiles:
         if self._zn_files is None:
-            self._zn_files = ZnFiles(node_name=self.name)
+            self._zn_files = ZnFiles(node_name=self.stage_name)
         return self._zn_files
 
-    def pre_init(self):
-        """Function to be called prior to the init"""
+    def pre_init(self, name: str, load: bool, has_metadata: bool):
```
Comment (SamTov) [inline zntrack/core/zntrack.py:100]: `has_metadata` has no docstring

**Review by SamTov — APPROVED** [review]
It all looks good I just had a question about one method. We should also run the license header updater on this soon.

**Comment by PythonFZ** [comment]
> It all looks good I just had a question about one method. We should also run the license header updater on this soon.

Can you make the license header an issue? Which files are missing?

**Comment by SamTov** [comment]
> > It all looks good I just had a question about one method. We should also run the license header updater on this soon.
> 
> Can you make the license header an issue? Which files are missing?

No files are missing all of the headers are just the old ones.

**Pull request merged**